// customer-service/src/app.controller.ts
import { Controller, Get } from '@nestjs/common';
import { AppService } from './app.service';
import { MessagePattern } from '@nestjs/microservices';
import { User } from './schemas/user.schema';

@Controller()
export class AppController {
  constructor(private readonly appService: AppService) {}
  @MessagePattern({ cmd: 'getUserById' })
  async getUserById(data: { id: string }): Promise<User> {
    try {
    } catch (err) {
      throw new Error(err);
    }
    const result = await this.appService.getUserById(data.id);
    return result;
  }
  @MessagePattern({ cmd: 'getUserByUsername' })
  async getUserByUsername(data: { username: string }): Promise<User> {
    try {
      const result = await this.appService.getUserByUsername(data.username);
      return result;
    } catch (err) {
      throw new Error(err);
    }
  }
  @MessagePattern({ cmd: 'getUserByEmail' })
  async getUserByEmail(data: { email_address: string }): Promise<any> {
    try {
      const result = await this.appService.getUserByEmail(data.email_address);
      return result;
    } catch (err) {
      throw new Error(err);
    }
  }
  @MessagePattern({ cmd: 'checkPhoneExist' })
  async checkPhoneExist(data: { phone_number: string }): Promise<Boolean> {
    try {
      const result = await this.appService.checkPhoneExist(data.phone_number);
      return result;
    } catch (err) {
      throw new Error(err);
    }
  }
  @MessagePattern({ cmd: 'createUser' })
  async createUser(data: { user: User }): Promise<User> {
    try {
      const result = await this.appService.createUser(data.user);
      return result;
    } catch (err) {
      throw new Error(err);
    }
  }
}
