// auth-service/src/app.service.ts
import { Inject, Injectable, NotFoundException } from '@nestjs/common';
import { LoginDto } from './dtos/login.dto';
import { ClientProxy } from '@nestjs/microservices';
import { RegisterDto } from './dtos/register.dto';
import { lastValueFrom } from 'rxjs';
import { bcrypt } from 'bcrypt';
@Injectable()
export class AppService {
  constructor(
    @Inject('CUSTOMER_SERVICE') private customerClient: ClientProxy,
  ) {}
  async login(data: LoginDto): Promise<any> {
    try {
      const exist_user = await this.customerClient.send(
        {
          cmd: 'getUserByUsername',
        },
        { username: data.username },
      );
      if (!exist_user) {
        throw new NotFoundException('Không tìm thấy user');
      }
    } catch (err) {
      throw new Error(err);
    }
  }
  async register(data: RegisterDto): Promise<any> {
    try {
      const exist_phone = await lastValueFrom(
        this.customerClient.send(
          { cmd: 'checkPhoneExist' },
          { phone_number: data.phone_number },
        ),
      );
      const exist_email = await lastValueFrom(
        this.customerClient.send(
          { cmd: 'getUserByEmail' },
          { email: data.email_address },
        ),
      );
      if (exist_email == true) {
        throw new Error(
          'Email đã được đăng kí bằng 1 tài khoản khác. Vui lòng sử dụng email khác',
        );
      }
      if (exist_phone == true) {
        throw new Error('Đăng kí thất bại ! Số điện thoại đã tồn tại');
      }

      const salt = await bcrypt.genSalt(10);
      const hashPass = await bcrypt.hash(data.password, salt);
      const newUser = {
        fullname: data.fullname,
        gender: data.gender,
        username: data.username,
        password: hashPass,
        dob: data.dob,
        avatar_url: data.avatar_url,

        email: {
          email_address: data.email_address,
          verified: false,
        },
        phone: {
          phone_number: data.phone_number,
          verified: false,
        },

        is_active: true,
      };
      const savedUser = await lastValueFrom(
        this.customerClient.send({ cmd: 'createUser' }, { user: newUser }),
      );
    } catch (err) {
      throw new Error(err);
    }
  }
}
